dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
-                else
+                catch (Exception ex) when (!ex.IsCriticalException())
                 {
-                    this._logger?.LogInformation("Action: No action to take");
+                    nextStep.Observation = $"Error invoking action {nextStep.Action} : {ex.Message}";
+                    this._logger?.LogWarning(ex, "Error invoking action {Action}", nextStep.Action);
                 }
 
-                // sleep 3 seconds
-                await Task.Delay(this.Config.MinIterationTimeMs).ConfigureAwait(false);
+                this._logger?.LogInformation("Observation: {Observation}", nextStep.Observation);
+                chatHistory?.AddUserMessage($"{Observation} {nextStep.Observation}");
+            }
+            else
+            {
+                this._logger?.LogInformation("Action: No action to take");
+                if (!string.IsNullOrEmpty(nextStep.Thought))
+                {
+                    chatHistory?.AddAssistantMessage($"{Thought} {nextStep.Thought}");
+                }
             }
 
-            context.Variables.Update($"Result not found, review _stepsTaken to see what happened.\n{JsonSerializer.Serialize(stepsTaken)}");
+            // sleep for a bit to avoid rate limiting
+            await Task.Delay(this.Config.MinIterationTimeMs, token).ConfigureAwait(false);
+        }
+
+        AddExecutionStatsToContext(stepsTaken, context, this.Config.MaxIterations);
+        context.Variables.Update($"Result not found, review _stepsTaken to see what happened.\n{JsonSerializer.Serialize(stepsTaken)}");
+
+        return context;
+    }
+
+    #region setup helpers
+    // Function to load PromptConfig from embedded resource
+    private static PromptTemplateConfig LoadPromptConfigFromResource()
+    {
+        string promptConfigString = EmbeddedResource.Read("Skills.StepwiseStep.config.json");
+        return !string.IsNullOrEmpty(promptConfigString) ? PromptTemplateConfig.FromJson(promptConfigString) : new PromptTemplateConfig();
+    }
+
+    private async Task<ChatHistory> InitializeChatHistoryAsync(ChatHistory chatHistory, IAIService aiService, SKContext context)
+    {
+        // this.CreateChatHistory(this._kernel, out var aiService);
+        var chatCompletion = aiService as IChatCompletion;
+        var textCompletion = aiService as ITextCompletion;
+
+        string userManual = await this.GetUserManualAsync(context).ConfigureAwait(false);
+        string userQuestion = await this.GetUserQuestionAsync(context).ConfigureAwait(false);
+
+#pragma warning disable CA2016 // Forward the 'CancellationToken' parameter to methods that take one -- the method is obsolete and to be removed.
+        var systemContext = this._kernel.CreateNewContext();
+#pragma warning restore CA2016
+
+        systemContext.Variables.Set("suffix", this.Config.Suffix);
+
+        if (chatCompletion is null)
+        {
+            systemContext.Variables.Set("functionDescriptions", userManual);
+        }
```

Comment:
I don't understand what this is doing. We overwrite the function descriptions with the userManual. Why?

Reply:
Will add explanation, cause it does appear confusing -- and may be worth changing.

In short -- I found when using Chat Completion, that putting the available functions as the first user message, instead of injected into the system message, to have better results (the LLM would respect it and hallucinate less leading to fewer iterations to answer asks). The more I learn about function calling support, and the fact it injects into the system message behind the scenes, makes me question my anecdotal experience.

Reply:
Added comment. Going to run a comp right now. Appreciate folks thoughts on default behavior.

Reply:
Okay, yeah, I'm not seeing that difference in these examples. I'm going to revert this special logic.